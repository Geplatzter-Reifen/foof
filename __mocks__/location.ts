type StageMockConstructorArgs = {
  latitude: number;
  longitude: number;
  recordedAt?: number;
  stage: { id: string; name: string };
  collections: any;
};

export class MockStage {
  constructor({
    latitude = 50.0826,
    longitude = 8.24,
    recordedAt = new Date("2024-12-03 12:00").getTime(),
    stage = { id: "mock-stage-id", name: "Mock Stage" },
    collections = {},
  }: Partial<StageMockConstructorArgs> = {}) {
    this.latitude = latitude;
    this.longitude = longitude;
    this.recordedAt = recordedAt;
    this.stage = stage;
    this.collections = collections;
  }
  static table = "stages";

  static associations = {
    tours: { type: "belongs_to", key: "tour_id" },
    locations: { type: "has_many", foreignKey: "stage_id" },
  };

  // Properties
  latitude: number;
  longitude: number;
  recordedAt?: number;
  stage;
  collections;
}
