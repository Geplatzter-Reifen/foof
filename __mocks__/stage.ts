type StageMockConstructorArgs = {
  title: string;
  isActive: boolean;
  startedAt: number;
  finishedAt: number | undefined;
  distance: number;
  avgSpeed: number;
  tour: { id: string; name: string };
  locations: [{ id: string; latitude: number; longitude: number }];
  collections: any;
};

export class MockStage {
  constructor({
    title = "Mock Stage",
    isActive = true,
    startedAt = new Date("2024-12-03 12:00").getTime(),
    finishedAt = undefined,
    distance = 42.195,
    avgSpeed = 10.5,
    tour = { id: "mock-tour-id", name: "Mock Tour" },
    locations = [{ id: "mock-location-id", latitude: 0, longitude: 0 }],
    collections = {},
  }: Partial<StageMockConstructorArgs> = {}) {
    this.title = title;
    this.isActive = isActive;
    this.startedAt = startedAt;
    this.finishedAt = finishedAt;
    this.distance = distance;
    this.avgSpeed = avgSpeed;
    this.tour = tour;
    this.collections = collections;
    this.locations = locations;
  }
  static table = "stages";

  static associations = {
    tours: { type: "belongs_to", key: "tour_id" },
    locations: { type: "has_many", foreignKey: "stage_id" },
  };

  // Properties
  title: string;
  isActive: boolean;
  startedAt: number;
  finishedAt?: number;
  distance: number;
  avgSpeed: number;
  tour: { id: string; name: string };
  locations: [{ id: string; latitude: number; longitude: number }];
  collections;
}
