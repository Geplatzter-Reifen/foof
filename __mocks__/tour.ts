type TourMockConstructorArgs = {
  title: string;
  isActive: boolean;
  startedAt?: number;
  finishedAt?: number;
  routes: { id: string; name: string };
  collections: any;
};

export class MockStage {
  constructor({
    title = "Mock Title",
    isActive = true,
    startedAt = new Date("2024-12-03 12:00").getTime(),
    finishedAt = undefined,
    routes = { id: "mock-route-id", name: "Mock Route" },
    collections = {},
  }: Partial<TourMockConstructorArgs> = {}) {
    this.title = title;
    this.isActive = isActive;
    this.startedAt = startedAt;
    this.finishedAt = finishedAt;
    this.routes = routes;
    this.collections = collections;
  }
  static table = "tours"; // bind the model to specific table
  static associations = {
    stages: { type: "has_many", foreignKey: "tour_id" },
    routes: { type: "has_many", foreignKey: "tour_id" },
  };

  // Properties
  title: string;
  isActive: boolean;
  startedAt: number;
  finishedAt?: number;
  routes;
  collections;
}
