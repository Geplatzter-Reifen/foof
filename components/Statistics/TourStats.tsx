import React from "react";
import { DimensionValue, StyleSheet, View } from "react-native";
import {
  formatDate,
  DateFormat,
  getTotalMillisecondsString,
} from "@/utils/dateUtil";
import {
  Icon,
  ProgressBar,
  Text,
  ThemeType,
  useTheme,
} from "@ui-kitten/components";
import { Tour, Stage } from "@/database/model/model";
import {
  getTourDuration,
  getTourDistance,
  getTourAverageSpeed,
} from "@/services/statisticsService";
import { withObservables } from "@nozbe/watermelondb/react";

let activeTour: Tour | undefined = undefined;
let tourStages: Stage[] = [];

function TourStats() {
  const theme = useTheme();
  const styles = makeStyles(theme);

  const progress = getTourDistance(tourStages) / 1001;

  return (
    <View style={styles.container}>
      <View style={styles.progressContainer}>
        <ProgressBar progress={progress} style={styles.progressBar} />
        <Text
          style={{
            position: "absolute",
            top: "28%",
            left:
              progress < 0.18
                ? (`${((progress + 0.04) * 100).toFixed(2)}%` as DimensionValue)
                : progress < 1
                  ? (`${((progress - 0.125) * 100).toFixed(2)}%` as DimensionValue)
                  : ("44%" as DimensionValue),
            color: progress >= 0.15 ? "#fff" : theme["text-basic-color"],
            fontSize: 19,
          }}
        >
          {(progress * 100).toFixed(1) + "%"}
        </Text>
      </View>
      <View style={styles.statsContainer}>
        <View style={styles.stat_column}>
          <View style={styles.stat_row}>
            <Icon name="arrows-left-right" style={styles.icon_style} />
            <Text>{getTourDistance(tourStages).toFixed(1) + " km"}</Text>
          </View>
          <View style={styles.stat_row}>
            <Icon name="arrow-up-right-dots" style={styles.icon_style} />
            <Text>{"0 m"}</Text>
          </View>
          <View style={styles.stat_row}>
            <Icon name="gauge-high" style={styles.icon_style} />
            <Text>{getTourAverageSpeed(tourStages).toFixed(1) + " km/h"}</Text>
          </View>
          <View style={styles.stat_row}>
            <Icon name="bolt" style={styles.icon_style} />
            <Text>{0 + " kcal"}</Text>
          </View>
        </View>
        <View style={styles.stat_column}>
          <View style={styles.stat_row}>
            <Icon name="calendar-plus" style={styles.icon_style} />
            <Text>
              {activeTour!.startedAt
                ? formatDate(activeTour!.startedAt, DateFormat.DATE)
                : "--"}
            </Text>
          </View>
          {activeTour!.finishedAt && (
            <View style={styles.stat_row}>
              <Icon name="calendar-check" style={styles.icon_style} />
              <Text>{formatDate(activeTour!.finishedAt, DateFormat.DATE)}</Text>
            </View>
          )}
          <View style={styles.stat_row}>
            <Icon name="clock" style={styles.icon_style} />
            <Text>
              {getTotalMillisecondsString(getTourDuration(tourStages))}
            </Text>
          </View>
        </View>
      </View>
    </View>
  );
}

// observe the stage (tracks updates to the stage)
const enhance = withObservables(["stage"], ({ stage }: { stage: Stage }) => ({
  stage,
}));

const EnhancedTourStats = enhance(TourStats);

// Bridge component that determines which TourStats component to render based on the active stage.
const Bridge = ({ stages }: { stages: Stage[] }) => {
  tourStages = stages;
  for (const stage of stages) {
    if (stage.isActive) {
      return <EnhancedTourStats stage={stage} />;
    }
  }
  return <TourStats />;
};

// observe stages of a tour (only tracks create and delete in the stages table)
const enhanceV2 = withObservables(["tour"], ({ tour }: { tour: Tour }) => {
  activeTour = tour;
  return {
    stages: tour.stages,
  };
});

const EnhancedTourStatsV2 = enhanceV2(Bridge);

export default EnhancedTourStatsV2;

const makeStyles = (theme: ThemeType) => {
  return StyleSheet.create({
    container: {
      backgroundColor: theme["color-primary-transparent-500"],
    },
    progressContainer: {
      padding: 8,
      margin: 10,
      marginBottom: 0,
    },
    progressBar: {
      height: 27,
      //@ts-ignore
      indicatorColor: theme["color-primary-500"],
      backgroundColor: theme["background-basic-color-2"],
      borderRadius: 6,
    },
    statsContainer: {
      flexDirection: "row",
      alignItems: "center",
      height: "auto",
      paddingHorizontal: 20,
      paddingBottom: 10,
    },
    stat_column: {
      flex: 1,
      alignItems: "flex-start",
      justifyContent: "flex-start",
    },
    stat_row: {
      flexDirection: "row",
      alignItems: "center",
    },
    icon_style: {
      alignSelf: "center",
      marginHorizontal: 4,
      height: 15,
      width: "100%",
      minWidth: 17,
    },
  });
};
