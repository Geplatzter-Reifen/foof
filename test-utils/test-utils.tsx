// https://testing-library.com/docs/react-testing-library/setup/#add-custom-queries
import React from "react";
import { render, RenderOptions } from "@testing-library/react-native";
import { FontAwesomeIconsPack } from "@/components/Font/fontAwesome";
import { mapping } from "@eva-design/eva";
import {
  IconRegistry,
  ApplicationProvider,
  ModalService,
} from "@ui-kitten/components";
import * as eva from "@eva-design/eva";
import { foofDarkTheme, foofLightTheme } from "@/constants/custom-theme";
import { library } from "@fortawesome/fontawesome-svg-core";
import { far } from "@fortawesome/free-regular-svg-icons";
import { fas } from "@fortawesome/free-solid-svg-icons";
import { fab } from "@fortawesome/free-brands-svg-icons";

// Iconpacks laden
library.add(far, fas, fab);

// ModalService wird verwendet, um die Modal-Positionierung zu steuern
ModalService.setShouldUseTopInsets = true;

type Props = {
  children: React.ReactNode;
  use_dark_theme?: boolean;
};
// Der Wrapper wird verwendet, um die Provider zu umschließen, die für die Komponenten notwendig sind
const AllTheProviders = ({ children, use_dark_theme = true }: Props) => {
  const theme = use_dark_theme
    ? { ...eva.dark, ...foofDarkTheme }
    : { ...eva.light, ...foofLightTheme };
  return (
    <>
      <IconRegistry icons={FontAwesomeIconsPack} />
      <ApplicationProvider {...eva} customMapping={mapping} theme={theme}>
        {children}
      </ApplicationProvider>
    </>
  );
};

// Erstellt einen Custom-Render, der die Provider um die Komponenten wickelt
const customRender = (
  component: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => {
  return render(component, { wrapper: AllTheProviders, ...options });
};

// Exportiert alle Funktionen der Testing Library, weswegen man diese dann auch von dieser Datei importieren kann
export * from "@testing-library/react-native";
export { customRender as render };
