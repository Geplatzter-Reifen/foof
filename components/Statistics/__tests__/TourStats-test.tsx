import { MockTour } from "@/__mocks__/tour";
import TourStats from "../TourStats";
import { render, waitFor, act } from "@/test-utils/test-utils";

describe("TourStats", () => {
  beforeEach(() => {
    jest
      .spyOn(global.Date, "now")
      .mockImplementation(() => new Date("2024-12-17 10:00").getTime());
  });

  it("should render correctly", async () => {
    const tour = new MockTour();

    let view;
    // Hier wird eine Warning ausgegeben, dass soll aber anscheinend so
    await act(async () => {
      // @ts-ignore
      view = await waitFor(() => render(<TourStats tour={tour} />));
    });
    expect(view).toMatchSnapshot();
  });
  //TODO: Add more tests
});
