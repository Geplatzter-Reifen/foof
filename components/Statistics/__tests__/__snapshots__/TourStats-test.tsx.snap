// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TourStats should render correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(51, 102, 255, 0.40)",
      "flexDirection": "row",
      "height": "auto",
      "paddingHorizontal": 20,
      "paddingVertical": 5,
    }
  }
>
  <View
    style={
      {
        "alignItems": "flex-start",
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "transform": [
              {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text />
      </View>
      <Text
        appearance="default"
        category="p1"
        status="basic"
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
            },
            undefined,
          ]
        }
      >
        03.12.2024
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "transform": [
              {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text />
      </View>
      <Text
        appearance="default"
        category="p1"
        status="basic"
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
            },
            undefined,
          ]
        }
      >
        00:00 h
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "flex-start",
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "transform": [
              {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text />
      </View>
      <Text
        appearance="default"
        category="p1"
        status="basic"
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
            },
            undefined,
          ]
        }
      >
        0.00 km
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "transform": [
              {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text />
      </View>
      <Text
        appearance="default"
        category="p1"
        status="basic"
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
            },
            undefined,
          ]
        }
      >
        0 m
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "transform": [
              {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text />
      </View>
      <Text
        appearance="default"
        category="p1"
        status="basic"
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
            },
            undefined,
          ]
        }
      >
        0.0 km/h
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "transform": [
              {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text />
      </View>
      <Text
        appearance="default"
        category="p1"
        status="basic"
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
            },
            undefined,
          ]
        }
      >
        0 kcal
      </Text>
    </View>
  </View>
</View>
`;
